src/senaite/patient/api.py

```diff
@@ -202,67 +205,80 @@ def to_datetime(date_value, default=None, tzinfo=None):
+def to_ymd(val, default=_marker):
     """Returns a representation of a relative delta in ymd format
     """
-    if not isinstance(delta, relativedelta):
-        raise TypeError("delta parameter must be a relative_delta")
+    if not isinstance(val, relativedelta):
+        if is_ymd(val):
+            return val
+        if default is _marker:
+            raise TypeError("delta parameter must be a relative_delta or ymd")
+        return default
 
     ymd = list("ymd")
-    diff = map(str, (delta.years, delta.months, delta.days))
+    diff = map(str, (val.years, val.months, val.days))
     age = filter(lambda it: int(it[0]), zip(diff, ymd))
     return " ".join(map("".join, age))
 
 
 def is_ymd(ymd):
     """Returns whether the string represents a period in ymd format
     """
-    valid = map(lambda p: p in ymd, "ymd")
-    return any(valid)
+    values = get_years_months_days(ymd, default=None)
+    return values is not None
 
 
-def get_birth_date(age_ymd, on_date=None):
-    """Returns the birth date given an age in ymd format and the date when age
-    was recorded or current datetime if None
+def get_years_months_days(ymd, default=_marker):
+    """Returns a tuple of (years, month, days)
     """
-    on_date = to_datetime(on_date, default=datetime.now())
-
     def extract_period(val, period):
-        num = re.findall(r'(\d{1,2})'+period, val) or [0]
+        num = re.findall(r'(\d{1,2})'+period, str(val)) or [0]
         return api.to_int(num[0], default=0)
 
-    # Extract the periods
-    years = extract_period(age_ymd, "y")
-    months = extract_period(age_ymd, "m")
-    days = extract_period(age_ymd, "d")
+    years = extract_period(ymd, "y")
+    months = extract_period(ymd, "m")
+    days = extract_period(ymd, "d")
     if not any([years, months, days]):
-        raise AttributeError("No valid ymd: {}".format(age_ymd))
+        if default is _marker:
+            raise AttributeError("No valid ymd: {}".format(ymd))
+        return default
 
+    return years, months, days
+
+
+def get_birth_date(age_ymd, on_date=None, default=_marker):
+    """Returns the birth date given an age in ymd format and the date when age
+    was recorded or current datetime if None
+    """
+    try:
+        ymd = to_ymd(age_ymd)
+        years, months, days = get_years_months_days(ymd)
+    except AttributeError:
+        if default is _marker:
+            raise AttributeError("No valid ymd: {}".format(age_ymd))
+        return default
+
+    on_date = dtime.to_dt(on_date) or datetime.now()
     dob = on_date - relativedelta(years=years, months=months, days=days)
     return dob
 
 
 def get_age_ymd(birth_date, on_date=None):
     """Returns the age at on_date if not None. Otherwise, current age
     """
-    delta = get_relative_delta(birth_date, on_date)
-    return to_ymd(delta)
+    try:
+        delta = dtime.get_relative_delta(birth_date, on_date)
+        return to_ymd(delta)
+    except (ValueError, TypeError):
+        return None
```

Comment:
You do everywhere the `default` handling except of here you return `None`? What happened? 😅

Maybe you can skip here the `ValueError`. At least I see only the explicit `TypeError` that is raised by `to_ymd`.

Reply:
Done